compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1092,6 +1093,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
                     val tyconIsInjective =
                       (tycon1sym.isClass || tycon2sym.isClass)
                       && (!touchedGADTs || gadtIsInstantiated)
+                      && !frozenGadt
```

Comment:
Note that this is a workaround for the frozenGadt problem fixed by #13380. So maybe we should first merge #13380, remove this line and see if any test breaks, and then merge this PR.